File: android-app/app/src/main/java/io/fxguard/kaya/data/realtime/SocketService.kt
/**
 * Kaya CRM — Socket.IO staff channel
 * @file    android-app/.../data/realtime/SocketService.kt
 * @layer   android
 * @owner   Ersan Jahed Tabrizi <ersanjahedtabrizi@gmail.com>
 * @see     docs/MOBILE-APP.md
 */
package io.fxguard.kaya.data.realtime

import io.fxguard.kaya.data.preferences.SessionStore
import io.socket.client.IO
import io.socket.client.Socket
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import org.json.JSONObject
import java.net.URI

data class SocketEvent(
    val name: String,
    val conversationId: String? = null,
    val threadId: String? = null,
    val ticketId: String? = null,
    val title: String? = null,
    val body: String? = null,
    val silent: Boolean = false,
)

class SocketService(private val session: SessionStore) {
    private var socket: Socket? = null
    private val _events = MutableSharedFlow<SocketEvent>(extraBufferCapacity = 64)
    val events: SharedFlow<SocketEvent> = _events

    @Synchronized
    fun connect() {
        disconnect()
        val token = session.token ?: return
        val opts = IO.Options().apply {
            auth = mapOf("token" to token)
            extraHeaders = mapOf("Authorization" to listOf("Bearer $token"))
            query = "token=$token"
            reconnection = true
            transports = arrayOf("websocket", "polling")
        }
        val s = IO.socket(URI.create(session.baseUrl), opts)
        fun listen(name: String) {
            s.on(name) { args -> _events.tryEmit(parseEvent(name, args)) }
        }
        listen("new_message")
        listen("message_sent")
        listen("assigned_message")
        listen("ticket_assigned")
        listen("ticket_reply_notification")
        listen("task_assigned")
        listen("internal_message")
        listen("internal_thread_updated")
        listen("important_announcement")
        listen("call_invite")
        listen("call_offer")
        runCatching { s.connect() }
        socket = s
    }

    @Synchronized
    fun disconnect() {
        socket?.off()
        socket?.disconnect()
        socket = null
    }

    fun emitRead(conversationId: String) {
        val s = socket ?: return
        runCatching {
            s.emit("join_conversation", JSONObject().put("conversationId", conversationId))
        }
    }

    private fun parseEvent(name: String, args: Array<Any>): SocketEvent {
        val obj = args.firstOrNull() as? JSONObject
        val message = obj?.optJSONObject("message")
        val customer = obj?.optJSONObject("customer")
        val fromUser = obj?.optJSONObject("fromUser") ?: message?.optJSONObject("fromUser")
        val direction = message?.optString("direction").orEmpty()
        val silent = name == "message_sent" || name == "internal_thread_updated" || direction == "outgoing"
        val conversationId = obj?.optString("conversationId")?.ifBlank {
            obj.optJSONObject("conversation")?.optString("id")
        }?.ifBlank { null }
        val threadId = obj?.optString("threadId")?.ifBlank { null }
        val ticketId = obj?.optString("ticketId")?.ifBlank { null }
        val msgType = message?.optString("type").orEmpty()
        val rawBody = listOf(
            message?.optString("content"),
            message?.optString("body"),
        ).firstOrNull { !it.isNullOrBlank() }.orEmpty()
        val mediaBody = when {
            rawBody.isNotBlank() -> rawBody
            msgType == "image" -> "📷"
            msgType == "video" -> "🎬"
            msgType == "audio" || msgType == "ptt" || msgType == "voice" -> "🎤"
            msgType == "sticker" || msgType == "document" -> "📎"
            else -> ""
        }
        val title = when (name) {
            "new_message", "assigned_message" ->
                customer?.optString("name")?.ifBlank { null }
                    ?: obj?.optString("customerName")?.ifBlank { null }
            "internal_message" -> fromUser?.optString("name")?.ifBlank { null }
            "important_announcement" -> obj?.optString("title")?.ifBlank { null }
            "ticket_assigned", "ticket_reply_notification" ->
                obj?.optString("title")?.ifBlank { obj.optString("ticketNumber") }?.ifBlank { null }
            "task_assigned" -> obj?.optString("title")?.ifBlank { null }
            "call_invite", "call_offer" -> obj?.optString("fromUserName")?.ifBlank { null }
            else -> null
        }
        val body = when (name) {
            "new_message", "assigned_message", "internal_message" -> mediaBody.ifBlank { null }
            "important_announcement" -> obj?.optString("body")?.ifBlank { null }
            "ticket_reply_notification" -> obj?.optString("replyContent")?.ifBlank { null }
            "call_invite", "call_offer" ->
                if (obj?.optString("type") == "video") "video" else "voice"
            else -> obj?.optString("title")?.ifBlank { null }
        }
        return SocketEvent(
            name = name,
            conversationId = conversationId,
            threadId = threadId,
            ticketId = ticketId,
            title = title,
            body = body,
            silent = silent,
        )
    }
}

File: ios-app/KayaStaff/Core/SocketService.swift
/**
 * Kaya CRM — Socket.IO v4 staff channel (websocket + Bearer header, never query)
 * @file    ios-app/KayaStaff/Core/SocketService.swift
 * @layer   ios
 * @owner   Ersan Jahed Tabrizi <ersanjahedtabrizi@gmail.com>
 * @see     docs/MOBILE-APP.md
 */
import Foundation

struct SocketEvent: Sendable {
    var name: String
    var conversationId: String?
    var threadId: String?
    var ticketId: String?
    var title: String?
    var body: String?
    var silent: Bool
}

final class SocketService: NSObject, URLSessionWebSocketDelegate {
    private let session: SessionStore
    private var webSocket: URLSessionWebSocketTask?
    private var urlSession: URLSession?
    private var receiveLoop = false
    private var reconnectWorkItem: DispatchWorkItem?
    private let lock = NSLock()
    private(set) var isConnected = false
    var onEvent: ((SocketEvent) -> Void)?

    init(session: SessionStore) {
        self.session = session
        super.init()
    }

    func connect() {
        disconnect()
        guard let token = session.token, !token.isEmpty else { return }
        guard let ws = websocketURL() else { return }
        var req = URLRequest(url: ws)
        req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        req.timeoutInterval = 30
        let config = URLSessionConfiguration.default
        config.waitsForConnectivity = true
        let urlSession = URLSession(configuration: config, delegate: self, delegateQueue: nil)
        let task = urlSession.webSocketTask(with: req)
        lock.lock()
        self.urlSession = urlSession
        self.webSocket = task
        self.receiveLoop = true
        lock.unlock()
        task.resume()
        listen()
    }

    func disconnect() {
        reconnectWorkItem?.cancel()
        reconnectWorkItem = nil
        lock.lock()
        receiveLoop = false
        isConnected = false
        webSocket?.cancel(with: .goingAway, reason: nil)
        webSocket = nil
        urlSession?.invalidateAndCancel()
        urlSession = nil
        lock.unlock()
    }

    func emitRead(conversationId: String) {
        let payload: [String: Any] = ["conversationId": conversationId]
        guard JSONSerialization.isValidJSONObject(payload),
              let data = try? JSONSerialization.data(withJSONObject: payload),
              let json = String(data: data, encoding: .utf8)
        else { return }
        send("42[\"join_conversation\",\(json)]")
    }

    private func websocketURL() -> URL? {
        var raw = session.baseUrl.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        if raw.hasPrefix("https://") {
            raw = "wss://" + raw.dropFirst("https://".count)
        } else if raw.hasPrefix("http://") {
            raw = "ws://" + raw.dropFirst("http://".count)
        } else {
            return nil
        }
        return URL(string: raw + "/socket.io/?EIO=4&transport=websocket")
    }

    private func send(_ text: String) {
        lock.lock()
        let socket = webSocket
        lock.unlock()
        socket?.send(.string(text)) { _ in }
    }

    private func listen() {
        lock.lock()
        let socket = webSocket
        let keep = receiveLoop
        lock.unlock()
        guard keep, let socket else { return }
        socket.receive { [weak self] result in
            guard let self else { return }
            switch result {
            case .failure:
                self.scheduleReconnect()
            case .success(let message):
                let text: String?
                switch message {
                case .string(let s): text = s
                case .data(let d): text = String(data: d, encoding: .utf8)
                @unknown default: text = nil
                }
                if let text { self.handleEngine(text) }
                self.listen()
            }
        }
    }

    private func handleEngine(_ raw: String) {
        guard let type = raw.first else { return }
        let rest = String(raw.dropFirst())
        switch type {
        case "0":
            if let token = session.token, !token.isEmpty,
               let enc = try? JSONSerialization.data(withJSONObject: ["token": token]),
               let json = String(data: enc, encoding: .utf8)
            {
                send("40\(json)")
            } else {
                send("40")
            }
        case "2":
            send("3")
        case "4":
            handleSocketIO(rest)
        default:
            break
        }
    }

    private func handleSocketIO(_ raw: String) {
        guard let sioType = raw.first else { return }
        let rest = String(raw.dropFirst())
        switch sioType {
        case "0":
            lock.lock()
            isConnected = true
            lock.unlock()
        case "1":
            scheduleReconnect()
        case "2":
            parseEventPacket(rest)
        case "4":
            scheduleReconnect()
        default:
            break
        }
    }

    private func parseEventPacket(_ raw: String) {
        var body = raw
        if let comma = body.firstIndex(of: ","), body.hasPrefix("/") {
            body = String(body[body.index(after: comma)...])
        }
        guard let data = body.data(using: .utf8),
              let arr = try? JSONSerialization.jsonObject(with: data) as? [Any],
              let name = arr.first as? String
        else { return }
        let payload = arr.count > 1 ? arr[1] as? [String: Any] : nil
        let event = Self.parseEvent(name: name, obj: payload ?? [:])
        DispatchQueue.main.async { [weak self] in
            self?.onEvent?(event)
        }
    }

    private func scheduleReconnect() {
        lock.lock()
        isConnected = false
        let should = receiveLoop
        lock.unlock()
        guard should, session.isLoggedIn else { return }
        reconnectWorkItem?.cancel()
        let work = DispatchWorkItem { [weak self] in self?.connect() }
        reconnectWorkItem = work
        DispatchQueue.main.asyncAfter(deadline: .now() + 2.5, execute: work)
    }

    func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask, didOpenWithProtocol protocol: String?) {
        lock.lock()
        isConnected = true
        lock.unlock()
    }

    func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
        if error != nil { scheduleReconnect() }
    }

    private static func parseEvent(name: String, obj: [String: Any]) -> SocketEvent {
        let message = obj["message"] as? [String: Any]
        let customer = obj["customer"] as? [String: Any]
        let fromUser = (obj["fromUser"] as? [String: Any]) ?? (message?["fromUser"] as? [String: Any])
        let direction = (message?["direction"] as? String) ?? ""
        let silent = name == "message_sent" || name == "internal_thread_updated" || direction == "outgoing"
        var conversationId = (obj["conversationId"] as? String)?.nilIfBlank
        if conversationId == nil {
            conversationId = ((obj["conversation"] as? [String: Any])?["id"] as? String)?.nilIfBlank
        }
        let threadId = (obj["threadId"] as? String)?.nilIfBlank
        let ticketId = (obj["ticketId"] as? String)?.nilIfBlank
        let msgType = (message?["type"] as? String) ?? ""
        let rawBody = [(message?["content"] as? String), (message?["body"] as? String)]
            .compactMap { $0?.nilIfBlank }
            .first ?? ""
        let mediaBody: String = {
            if !rawBody.isEmpty { return rawBody }
            switch msgType {
            case "image": return "📷"
            case "video": return "🎬"
            case "audio", "ptt", "voice": return "🎤"
            case "sticker", "document": return "📎"
            default: return ""
            }
        }()
        let title: String? = {
            switch name {
            case "new_message", "assigned_message":
                return (customer?["name"] as? String)?.nilIfBlank ?? (obj["customerName"] as? String)?.nilIfBlank
            case "internal_message":
                return (fromUser?["name"] as? String)?.nilIfBlank
            case "important_announcement":
                return (obj["title"] as? String)?.nilIfBlank
            case "ticket_assigned", "ticket_reply_notification":
                return (obj["title"] as? String)?.nilIfBlank ?? (obj["ticketNumber"] as? String)?.nilIfBlank
            case "task_assigned":
                return (obj["title"] as? String)?.nilIfBlank
            case "call_invite", "call_offer":
                return (obj["fromUserName"] as? String)?.nilIfBlank
            default:
                return nil
            }
        }()
        let body: String? = {
            switch name {
            case "new_message", "assigned_message", "internal_message":
                return mediaBody.nilIfBlank
            case "important_announcement":
                return (obj["body"] as? String)?.nilIfBlank
            case "ticket_reply_notification":
                return (obj["replyContent"] as? String)?.nilIfBlank
            case "call_invite", "call_offer":
                return (obj["type"] as? String) == "video" ? "video" : "voice"
            default:
                return (obj["title"] as? String)?.nilIfBlank
            }
        }()
        return SocketEvent(
            name: name,
            conversationId: conversationId,
            threadId: threadId,
            ticketId: ticketId,
            title: title,
            body: body,
            silent: silent
        )
    }
}

private extension String {
    var nilIfBlank: String? {
        let t = trimmingCharacters(in: .whitespacesAndNewlines)
        return t.isEmpty ? nil : t
    }
}
